File: backend/src/schemas/car.py
"""Car-related API schemas and typed payloads."""

from decimal import Decimal
from typing import TypedDict

from pydantic import BaseModel


class CarResponse(BaseModel):
    """Serialized car listing returned by API."""

    id: int
    make: str
    model: str
    year: int
    price: Decimal
    color: str
    source_url: str

    class Config:
        """Pydantic model configuration."""

        from_attributes = True


class CarFilter(BaseModel):
    """Filter options used by API and bot search."""

    make: str | None = None
    model: str | None = None
    color: str | None = None
    max_price: float | None = None
    min_year: int | None = None


class CarUpsertPayload(TypedDict):
    """Normalized payload used for database upsert operations."""

    make: str
    model: str
    year: int
    price: float
    color: str
    source_url: str

File: frontend/src/types/car.ts
export type Car = {
  id: number;
  make: string;
  model: string;
  year: number;
  price: string;
  color: string;
  source_url: string;
};

export type CarFilters = {
  make?: string;
  model?: string;
  color?: string;
  max_price?: number;
  min_year?: number;
  limit?: number;
  offset?: number;
};
